### PR TITLE
Package rewrite

### PR DESCRIPTION
**Pull request by JSchoenbachler**

### CODE REVIEW

File: R/metapredict_confusion.R [hunk 1]
```diff
@@ -46,36 +53,42 @@ calcConfusionCv = function(cvFit, lambda, ematMerged, sampleMetadata,
...
 #' @export
 calcConfusionValidation = function(predsList, lambda, sampleMetadata,
-                                   className='class', classLevels=NA,
-                                   each=TRUE) {
+                                   className = 'class', classLevels = NA,
+                                   each = TRUE) {
   if (is.na(classLevels[1])) {
     classLevels = colnames(predsList[[1]])}
 
   if (each) {
```
Comment (jakejh) [inline R/metapredict_confusion.R]: `isTRUE(each)`

File: R/metapredict_confusion.R [hunk 2]
```diff
@@ -46,36 +53,42 @@ calcConfusionCv = function(cvFit, lambda, ematMerged, sampleMetadata,
...
     confusion = list()
     for (validationStudyName in names(predsList)) {
       predsProb = predsList[[validationStudyName]][,,1]
```
Comment (jakejh) [inline R/metapredict_confusion.R]: add space after commas

File: R/metapredict_merge.R [hunk 3]
```diff
@@ -1,16 +1,29 @@
...
+
+  order = if (is.unsorted(subStudyNames)) -1L else 1L
+  setorderv(metadataNow, 'study', order = order)
+
+  # mappingDf = metadataNow %>%
+  #   dplyr::group_by(!!matchSampleColname) %>%
+  #   dplyr::slice(1) %>%
+  #   dplyr::ungroup()
+
+  mappingDf = metadataNow[1,, by = ..matchSampleColname]
```
Comment (jakejh) [inline R/metapredict_merge.R]: Are you sure this does what you think? I would've done `metadataNow[, .SD[1L], by = ..matchSampleColname]`

File: R/metapredict_merge.R [hunk 4]
```diff
@@ -32,31 +45,38 @@ makeMatchSampleMapping = function(metadata, subStudyNames, matchSampleColname) {
...
   sampleMetadataList = list()
+  sampleMetadataAtomicDT = data.table(sampleMetadataAtomic)
 
   for (matchStudyName in unique(studyMetadataAtomic[[matchStudyColname]])) {
-    if (sum(studyMetadataAtomic[[matchStudyColname]]==matchStudyName)==1) {
+    if (sum(studyMetadataAtomic[[matchStudyColname]] == matchStudyName) == 1) {
       ematList[[matchStudyName]] = ematAtomicList[[matchStudyName]]
-      sampleMetadataList[[matchStudyName]] = dplyr::filter(sampleMetadataAtomic, study==matchStudyName)
+      # sampleMetadataList[[matchStudyName]] = dplyr::filter(sampleMetadataAtomic, study == matchStudyName)
+      sampleMetadataList[[matchStudyName]] = sampleMetadataAtomicDT[which(study == matchStudyName),]
```
Comment (jakejh) [inline R/metapredict_merge.R]: I expect to never again see `which` in the `i` argument of a data.table.

File: R/metapredict_merge.R [hunk 5]
```diff
@@ -32,31 +45,38 @@ makeMatchSampleMapping = function(metadata, subStudyNames, matchSampleColname) {
...
         edf = data.frame(rownames(ematAtomicList[[atomicStudyName]]), ematAtomicList[[atomicStudyName]])
         rownames(edf) = NULL
-        matchNames = tibble::tibble(sample = colnames(edf)[2:ncol(edf)]) %>%
-          dplyr::inner_join(sampleMetadataAtomic, by='sample') %>%
-          .[[matchSampleColname]]
+        # matchNames = tibble::tibble(sample = colnames(edf)[2:ncol(edf)]) %>%
+        #   dplyr::inner_join(sampleMetadataAtomic, by = 'sample') %>%
+        #   .[[matchSampleColname]]
+
+        matchNames = merge(data.table(sample = colnames(edf)[2:ncol(edf)]), sampleMetadataAtomicDT, by = 'sample', sort = FALSE)[[matchSampleColname]]
```
Comment (jakejh) [inline R/metapredict_merge.R]: Seems like you could use the new `mergeDataTable` function.

File: R/metapredict_merge.R [hunk 6]
```diff
@@ -32,31 +45,38 @@ makeMatchSampleMapping = function(metadata, subStudyNames, matchSampleColname) {
...
-      edfMerged = suppressWarnings(dplyr::bind_rows(edfListNow)) %>%
-        dplyr::group_by(geneId) %>%
-        dplyr::summarize_all(mergeFunc) %>%
-        data.frame(check.names=FALSE)
+      # edfMerged = suppressWarnings(dplyr::bind_rows(edfListNow)) %>%
+      #   dplyr::group_by(geneId) %>%
+      #   dplyr::summarize_all(mergeFunc) %>%
+      #   data.frame(check.names = FALSE)
+
+      edfMerged = data.frame(data.table(rbind(edfListNow)[[1]])[, by = geneId, lapply(.SD, mergeFunc)], check.names = FALSE)
```
Comment (jakejh) [inline R/metapredict_merge.R]: revise order of arguments in data.table to i, j, by. in general, never pass unnamed arguments after named arguments.

File: R/metapredict_merge.R [hunk 7]
```diff
@@ -105,33 +129,36 @@ mergeMatchStudyData = function(ematAtomicList, studyMetadataAtomic, matchStudyCo
...
     # ComBat can fail catastrophically
     ematListScaled = lapply(ematList2, function(emat) (emat - mean(emat)) / stats::sd(emat))
     ematMerged = do.call(cbind, ematListScaled)
-    sm = tibble::tibble(sample = colnames(ematMerged)) %>%
-      dplyr::inner_join(sampleMetadata, by='sample')
+
+    # sm = tibble::tibble(sample = colnames(ematMerged)) %>%
+    #   dplyr::inner_join(sampleMetadata, by = 'sample')
+
+    sm = merge(data.table(sample = colnames(ematMerged)), sampleMetadata, by = 'sample', sort = FALSE)
```
Comment (jakejh) [inline R/metapredict_merge.R]: use new function

File: R/metapredict_plot.R [hunk 8]
```diff
@@ -97,19 +98,21 @@ plotExpressionHeatmap = function(fitResult, lambda, ematMerged, sampleMetadata,
     emat = emat[,co$order]
   } else {
     if (is.na(classLevels[1])) {
-      sm = dplyr::filter(sampleMetadata, sample %in% colnames(ematMerged))
+      # sm = dplyr::filter(sampleMetadata, sample %in% colnames(ematMerged))
+      sm = data.table(sampleMetadata)[which(sample %in% colnames(ematMerged)),]
```
Comment (jakejh) [inline R/metapredict_plot.R]: You know what.

File: R/metapredict_plot.R [hunk 9]
```diff
@@ -97,19 +98,21 @@ plotExpressionHeatmap = function(fitResult, lambda, ematMerged, sampleMetadata,
...
       classLevels = unique(sm[[className]])}
 
-    ematSmallList = foreach(classLevel=classLevels) %do% {
+    # ematSmallList = foreach(classLevel=classLevels) %do% {
+    emat = foreach(classLevel=classLevels, .combine = cbind) %do% {
```
Comment (jakejh) [inline R/metapredict_plot.R]: whitespace

File: R/metapredict_plot.R [hunk 10]
```diff
@@ -126,10 +129,12 @@ plotExpressionHeatmap = function(fitResult, lambda, ematMerged, sampleMetadata,
...
-  annotation = tibble::tibble(sample = colnames(ematMerged)) %>%
-    dplyr::inner_join(sampleMetadata, by='sample') %>%
-    dplyr::select(!!names(annoLevels)) %>%
-    as.data.frame()
+  # annotation = tibble::tibble(sample = colnames(ematMerged)) %>%
+  #   dplyr::inner_join(sampleMetadata, by='sample') %>%
+  #   dplyr::select(!!names(annoLevels)) %>%
+  #   as.data.frame()
+  cols = names(annoLevels)
+  annotation = as.data.frame(merge(data.table(sample = colnames(ematMerged)), sampleMetadata, by = 'sample', sort = FALSE)[,..cols])
```
Comment (jakejh) [inline R/metapredict_plot.R]: use new function

File: R/metapredict_plot.R [hunk 11]
```diff
@@ -166,8 +171,9 @@ plotClassProbsCv = function(cvFit, lambda, ematMerged, sampleMetadata, className
                             size=1.5, ggplotArgs=NA) {
 
   sampleNames = colnames(ematMerged)
-  sm = tibble::tibble(sample = sampleNames) %>%
-    dplyr::inner_join(sampleMetadata, by='sample')
+  # sm = tibble::tibble(sample = sampleNames) %>%
+  #   dplyr::inner_join(sampleMetadata, by='sample')
+  sm = merge(data.table(sample = sampleNames), sampleMetadata, by = 'sample', sort = FALSE)
```
Comment (jakejh) [inline R/metapredict_plot.R]: use new function. you missed several of these.

File: R/metapredict_plot.R [hunk 12]
```diff
@@ -176,12 +182,17 @@ plotClassProbsCv = function(cvFit, lambda, ematMerged, sampleMetadata, className
   cvProbs = cvFit$fit.preval[,,which.min(abs(cvFit$lambda - lambda))]
   pList = list()
   for (studyName in studyNames) {
-    sampleNamesNow = dplyr::filter(sm, study==studyName)$sample
-    df = tibble::as_tibble(cvProbs[sm$study==studyName,])
+    # sampleNamesNow = dplyr::filter(sm, study==studyName)$sample
+    sampleNamesNow = sm[which(study == studyName), sample]
```
Comment (jakejh) [inline R/metapredict_plot.R]: fix

File: R/metapredict_plot.R [hunk 13]
```diff
@@ -192,7 +203,8 @@ plotClassProbsCv = function(cvFit, lambda, ematMerged, sampleMetadata, className
       if (any(df$trueClass==classLevel)) {
         idxTmp = c(idxTmp, 1:(sum(df$trueClass==classLevel)))}}
     df$idx = idxTmp
-    dfMolten = tidyr::gather(df, key='probClass', value='prob', !!classLevels)
+    # dfMolten = tidyr::gather(df, key='probClass', value='prob', !!classLevels)
+    dfMolten = melt(as.data.table(df), measure.vars = classLevels, variable.name = 'probClass', variable.factor = FALSE, value.name = 'prob')
```
Comment (jakejh) [inline R/metapredict_plot.R]: Unclear why code sometimes uses `as.data.table` and sometimes `data.table` when being passed a data.frame

File: R/metapredict_plot.R [hunk 14]
```diff
@@ -231,9 +243,12 @@ plotClassProbsValidation = function(predsList, sampleMetadata, className,
   pList = list()
   for (validationStudyName in names(predsList)) {
     df = data.frame(predsList[[validationStudyName]][,,1])
-    sm = tibble::tibble(sample = rownames(df)) %>%
-      dplyr::inner_join(sampleMetadata, by='sample')
-    df$study = sm$study
+    # sm = tibble::tibble(sample = rownames(df)) %>%
+    #   dplyr::inner_join(sampleMetadata, by='sample')
+
+    sm = merge(data.table(sample = rownames(df)), sampleMetadata, by = 'sample', sort = FALSE)
```
Comment (jakejh) [inline R/metapredict_plot.R]: fix

File: R/metapredict_predict.R [hunk 15]
```diff
@@ -107,26 +110,37 @@ metapredictCv = function(ematMerged, sampleMetadata, weights, alpha, nFolds=10,
...
-      as.matrix()
-    fitResult = glmnet::glmnet(t(ematMergedDisc), y, alpha=alpha, lambda=lambda,
-                               weights=weights[discoverySampleNames], standardize=FALSE, ...)
+
+    # y = tibble::tibble(sample = discoverySampleNames) %>%
+    #   dplyr::inner_join(sampleMetadata, by = 'sample') %>%
+    #   dplyr::select(!!className) %>%
+    #   as.matrix()
+
+    yDTM = data.table(sample = discoverySampleNames)[sampleMetadataDT, on = "sample", nomatch = 0]
```
Comment (jakejh) [inline R/metapredict_predict.R]: use consistent merge syntax

File: R/metapredict_predict.R [hunk 16]
```diff
@@ -171,14 +185,14 @@ makeCoefDf = function(fitResult, lambda, decreasing=TRUE, classLevels=NA) {
...
-    coefDf = Reduce(function(x, y) merge(x, y, by='geneId', all=TRUE), coefResultNonzero)
-    idx = do.call(order, c(coefDf[,2:ncol(coefDf)], list(decreasing=decreasing)))
+    coefDf = Reduce(function(x, y) merge(x, y, by = 'geneId', all = TRUE), coefResultNonzero)
+    idx = do.call(order, c(coefDf[,2:ncol(coefDf)], list(decreasing = decreasing)))
     coefDf = coefDf[idx,]
     coefDf[is.na(coefDf)] = 0
 
   } else {
-    coefDf = data.frame(names(coefResult[(coefResult@i)+1,]), coefResult[(coefResult@i)+1,], stringsAsFactors=FALSE)
+    coefDf = data.frame(names(coefResult[(coefResult@i)+1,]), coefResult[(coefResult@i)+1,], stringsAsFactors = FALSE)
```
Comment (jakejh) [inline R/metapredict_predict.R]: why does `coefDf` need to be a data.frame, especially if the rownames later get set to NULL?

File: R/metapredict_process.R [hunk 17]
```diff
@@ -73,58 +73,84 @@ fixGeoSampleNames = function(sampleNames) {
...
+  sampleNamesNew = gsub('\\.cel(\\.gz)*$', '', sampleNames, ignore.case = TRUE)
   return(sampleNamesNew)}
 
 
 getGeneProbeMappingAffy = function(mappingFilePath) {
-  mapping = utils::read.table(mappingFilePath, sep='\t', header=TRUE, stringsAsFactors=FALSE)
-  mappingUnique = unique(mapping[,c('Probe.Set.Name', 'Affy.Probe.Set.Name')])
-  mappingUnique = mappingUnique[apply(mappingUnique, MARGIN=1, function(r) !any(is.na(r))),]
+  # mapping = utils::read.table(mappingFilePath, sep = '\t', header = TRUE, stringsAsFactors = FALSE)
+  mapping = fread(mappingFilePath, sep = '\t', header = TRUE, stringsAsFactors = FALSE)
```
Comment (jakejh) [inline R/metapredict_process.R]: We talked about this.

File: R/metapredict_process.R [hunk 18]
```diff
@@ -73,58 +73,84 @@ fixGeoSampleNames = function(sampleNames) {
...
+  mappingUnique = unique(mapping[,c('geneId', 'probeSet')])
```
Comment (jakejh) [inline R/metapredict_process.R]: Actually, none of this seems to have changed from what we discussed on Friday. Is this your latest commit?

**Review by jakejh — CHANGES_REQUESTED** [review]
See various comments.

**Comment by JSchoenbachler** [comment]
I am reviewing your comments. I am going to delete and re-clone the repository locally from scratch, I think there is something weird going on. I'm not sure if while making edits I accidentally undid work or if this has something to do with my RStudio crashing, but I know I did a search and replace on both the "=" spacing issues and the "which" usage. I'm also seeing other things I know we talked about and changes I know I made (I did the fread change while we talked about it on the 1:1 call and pushed after making a few more changes).

I'm going to re-review the code once I've done that and go back through my notes to make sure each change I made is on the branch.

**Comment by jakejh** [comment]
Ok, now please go line by line to check every single thing.

**Comment by JSchoenbachler** [comment]
I've not reviewed fully yet, but I went to each place you had comments on and have done a quick review. I have everything you saw except for the makeCoefDf. Additionally, it looks like the `=` spacing was a problem in the `metapredict_plot` script was still there so I did a search and replace. I additionally searched all other files for any unspaced `=` and remaining instances of the `which` usage.

As I said, I'm not sure what exactly happened but I haven't had RStudio crash on me since re-cloning, so maybe it had something to do with that? I'm not sure, but I'm going to be reviewing in much more detail over then next bit.

File: R/metapredict_merge.R [hunk 19]
```diff
@@ -70,15 +90,19 @@ mergeMatchStudyData = function(ematAtomicList, studyMetadataAtomic, matchStudyCo
...
-    dplyr::select(!!colnamesKeep)
+  # studyMetadata = studyMetadataAtomic %>%
+  #   dplyr::group_by(!!matchStudyColname) %>%
+  #   dplyr::slice(1) %>%
+  #   dplyr::ungroup() %>%
+  #   # dplyr::mutate_(study = lazyeval::interp(~ c1, c1 = as.name(matchStudyColname))) %>%
+  #   dplyr::mutate(study = !!matchStudyColname) %>%
+  #   dplyr::select(!!colnamesKeep)
+
+  studyMetadata = studyMetadataAtomic[1,, by = ..matchStudyColname][,study := matchStudyColname][,..colnamesKeep]
```
Comment (jakejh) [inline R/metapredict_merge.R]: Same issue as above, I think this should use `SD[1L]`. You should test this.

**Comment by JSchoenbachler** [comment]
Sorry, accidentally closed instead of commenting.

I have re-requested your review with the latest commit.

**Comment by jakejh** [comment]
Are you completely confident I won't find anything we already went over?

**Comment by JSchoenbachler** [comment]
Yes.

**Comment by JSchoenbachler** [comment]
The only thing I will say I didn't specifically check or do while reviewing was the thing we discussed about breaking up single lines if they go above 80 characters.

File: DESCRIPTION [hunk 20]
```diff
@@ -21,24 +21,23 @@ Imports:
...
   cowplot (>= 0.9.3),
-  dplyr (>= 0.7.6),
+  data.table (>= 1.12.2),
   foreach (>= 1.4.3),
   GEOquery (>= 2.36.0),
   ggplot2 (>= 2.2.1),
   glmnet (>= 2.0-2),
   impute (>= 1.44.0),
+  iterators (>= 1.0.12),
   lazyeval (>= 0.2.1),
```
Comment (jakejh) [inline DESCRIPTION]: I believe lazyeval was necessary for dplyr-related stuff. Check if it can be removed.

File: R/metapredict_confusion.R [hunk 21]
```diff
@@ -1,3 +1,7 @@
+mergeDataTable = function(sampleList, sampleMetadata) {
+  merge(data.table(sample = sampleList), sampleMetadata, by = 'sample', sort = FALSE)
```
Comment (jakejh) [inline R/metapredict_confusion.R]: To make sure you internalize R's classes, I'm going to ask that you change the name of `sampleList`.

File: DESCRIPTION (hunk 20 above)
Comment (JSchoenbachler) [reply DESCRIPTION]: Done with latest commit.

File: R/metapredict_confusion.R (hunk 21 above)
Comment (JSchoenbachler) [reply R/metapredict_confusion.R]: Done with latest commit.

**Comment by jakejh** [comment]
Here are the places I think need another look. I want you to figure out the specific change that should be made. Most are related to whitespace, but not all, so be careful to make sure it's right.

- metapredict_merge.R: lines 59, 81, 101
- metapredict_plot.R: lines 24, 27, 31-32, 41, 45, 48 (use aes instead of aes_string), 89, 93, 103, 187, 189-208 (lots of code should change to make df a data.table from the start), 211 (use new `rows = ggplot2::vars` etc. syntax for `facet_grid`), 212 (use aes instead of aes_string), 252-268 (same comment as for lines 189-208), 271 (same comment as for 211), 273 (same comment as for 212)
- metapredict_predict.R: lines 124, 218, 219-220, 235, 236-237, 238, 240, remove or at least comment out the entire makeCoefDf function and its documentation
- metapredict_process.R: lines 81, 82, 83, 84-85, 93-94, 104-105, 122-125, 166, 216, 421, 443, 445
- should featureDf be called featureDt?
- example_script.R does not belong in the vignettes folder

**Comment by JSchoenbachler** [comment]
With the latest commits, I have everything except for the following:

metapredict_predict lines 206 & 235. This is how the code was set up previously and I'm not entirely sure how the `@` character works with this. Might need some help with this one or a nudge in the right direction.

metapredict_process line 414. Do you want to use a different `.combine` method or is there something else going on?

**Comment by jakejh** [comment]
metapredict_predict lines 206 and 235 are a simple issue of whitespace.

metapredict_process line 214 is indeed related to the .combine argument. What class of object does foreach return by default, and what class of object do we want?

**Comment by JSchoenbachler** [comment]
Re-requesting review.

**Comment by jakejh** [comment]
I thought we said the stuff with vignettes in CI and testing would stay separate until after we merged the rewrite into master.

**Comment by JSchoenbachler** [comment]
You're correct, I completely forgot to push that to a new branch while testing. I will roll that into a new branch and roll back here.

**Comment by jakejh** [comment]
Here are the places I think deserve another look:

- metapredict_plot: lines 23, 32, 45, 47 (spacing and use new syntax for `facet_wrap` based on `vars`), 89, 202, 211, 257, 263-265, 270
- metapredict_predict: lines 124, 222
- metapredict_process: 81, 90, 95, 159

**Comment by JSchoenbachler** [comment]
Latest push adresses all except:

metapredict_plot line 202
and
metapredict_process line 81

I'm not sure what you are wanting for these.

**Comment by jakejh** [comment]
What makes you think `names(cvFit$glmnet.fit$beta)` always has length 3? How does `fread` determine `sep`?

**Comment by JSchoenbachler** [comment]
Okay, changed in latest commit!

**Comment by jakejh** [comment]
- Line 47: What is the first argument for `facet_wrap`?
- Line 202 can be simplified further.
- Lines 211 and 270 should use a specific data.table function instead of do.call.
- Line 257 still has a problem, whose solution will affect line 263.
- Line 264 can be altered to remove the need for line 265.

**Comment by JSchoenbachler** [comment]
I believe I made the changes as you want them to be!

File: R/metapredict_merge.R [hunk 22]
```diff
@@ -32,33 +45,40 @@ makeMatchSampleMapping = function(metadata, subStudyNames, matchSampleColname) {
...
-      edfMerged = suppressWarnings(dplyr::bind_rows(edfListNow)) %>%
-        dplyr::group_by(geneId) %>%
-        dplyr::summarize_all(mergeFunc) %>%
-        data.frame(check.names=FALSE)
+      # edfMerged = suppressWarnings(dplyr::bind_rows(edfListNow)) %>%
+      #   dplyr::group_by(geneId) %>%
+      #   dplyr::summarize_all(mergeFunc) %>%
+      #   data.frame(check.names = FALSE)
+
+      edfMerged = data.frame(data.table(rbind(edfListNow)[[1]])[, lapply(.SD, mergeFunc)], by = geneId, check.names = FALSE)
```
Comment (jakejh) [inline R/metapredict_merge.R]: `by` is not an argument to `data.frame`, so this code should give an error. Have you tested this? I think the closing bracket is misplaced.

Comment (JSchoenbachler) [reply R/metapredict_merge.R]: I was testing each of the functions side by side with the dplyr ones, though its possible I mistakenly typed that or copied and pasted wrong when officially changing it. In any case, its been updated with the latest push.

**Comment by jakejh** [comment]
I went through the whole thing again, and found more issues not related to whitespace. Check my recent commits and let me know if you're ok with merging this bad boy into master.

**Comment by JSchoenbachler** [comment]
After reviewing many of the changes look good to me.

Was it really much slower using the data.table vectorization? When I was testing it was faster!

**Comment by jakejh** [comment]
The data.table version was about 3.5x slower on a few datasets that call `calcExprByGenes`, which two of the four datasets in the vignette do not.

**Pull request merged**